src/selector_gui.cpp

```diff
@@ -0,0 +1,386 @@
+ * This function is meant to be called from the parent's window's OnResize
+ * Event handler
+ *
+ * All of these parameters should be passed straight from the parent's window's OnResize
+ * Event handler, without any changes, or conditional checks
+ *
+ * @see Window::OnResize()
+ */
+void SelectorWidget::OnResize()
+{
+	this->vscroll->SetCapacityFromWidget(w, WID_SELECTOR_MATRIX);
+}
+
+/**
+ * Selector widget's OnEditboxChanged event handler
+ * This function is meant to be called from the parent's window's OnEditboxChanged
+ * Event handler
+ *
+ * @param wid The widget id of the editbox.
+ *
+ * All of these parameters should be passed straight from the parent's window's OnEditboxChanged
+ * Event handler, without any changes, or conditional checks
+ *
+ * @see Window::OnEditboxChanged()
+ */
+void SelectorWidget::OnEditboxChanged(WidgetID wid)
+{
+	if (wid != WID_SELECTOR_EDITBOX) return;
+
+	this->string_filter.SetFilterTerm(this->editbox.text.buf);
+	this->w->InvalidateData(0);
+	this->vscroll->SetCount(this->filtered_list.size());
+}
+
+/**
+ * Selector widget's DrawWidget method
+ * This function is meant to be called from the parent's window's DrawWidget
+ * method
+ *
+ * @param widget  Widget number.
+ * @param[In,out] size Size of the widget.
+ * @param padding Recommended amount of space between the widget content and the widget edge.
+ * @param[In,out] fill Fill step of the widget.
+ * @param[In,out] resize Resize step of the widget.
+ *
+ * All of these parameters should be passed straight from the parent's window's DrawWidget
+ * Event handler, without any changes, or conditional checks
+ *
+ * This function calls a profile defined function this->profile.DrawSection(),
+ * that actually draws the fields
+ *
+ * @see Window::DrawWidget()
+ * @see SelectorWidget::Profile
+ */
+void SelectorWidget::DrawWidget(const Rect &r, WidgetID widget)
+{
+	if (widget != WID_SELECTOR_MATRIX) return;
+
+	Rect line = r.WithHeight(row_height);
+
+	auto [first, last] = this->vscroll->GetVisibleRangeIterators(filtered_list);
+
+	for (auto it = first; it != last; ++it) {
+		const Rect ir = line.Shrink(WidgetDimensions::scaled.framerect);
+
+		if (this->shown[*it]) {
+			DrawFrameRect(line, COLOUR_BROWN, FR_LOWERED);
+		}
+		this->DrawSection(*it, ir.Shrink(WidgetDimensions::scaled.matrix));
+
+		line = line.Translate(0, row_height);
+	}
+}
+
+/**
+ * A function that updates and rebuilds the list of selectable items
+ * This function calls a profile defined function this->profile.RebuildList(),
+ * that actually rebuilds the list of items.
+ *
+ * It is also manages the shown items and the filtered_list lists
+ *
+ * @see SelectorWidget::Profile
+ */
+void SelectorWidget::RebuildList()
+{
+	this->list.clear();
+	this->filtered_list.clear();
+	this->PopulateList();
+	uint32_t max = 0;
+	for (const uint32_t id : list) {
+		if (id > max) {
+			max = id;
+		}
+	}
+	for (size_t i = this->shown.size(); i < max + 1; i++) {
+		shown.push_back(true);
+	}
```

Comment:
Why does shown treat the value of a list item as an index? If list contains "{0, 63}" then shown now contains everything from 0 to 63.

Also missing `this->` on the second `shown`.

Reply:
The shown list is the size of the max index, and it is so, because filtered_list is sometimes filtered, and it still needs to know if the item is shown or no to draw the depressed button

Reply:
I'm not really questioning what `shown` is for, more how it is implemented.

If only companies 1, 2 and 15 exist, then presumably shown has 15 elements. Maybe someone wants to make a graph by serviced industries, where there could be thousands of valid indices, but only a few might actually be relevant.

This could be something for a std::set (or std::unordered_set), which would allow only the required values to exist.

Reply:
Ok, I see that, and it will be a good change to do if such graph would ever exist, but it would be significantly slower rn, as we are comparing a simple indexing with a hash table lookup, and unless the memory footprint will really be an issue, (even thousands of bools is not that big), only then this should be changed. Of course if you say it should be changed, I will change it